builtin/game/misc.lua

```diff
@@ -4,74 +4,45 @@
 -- Misc. API functions
 --
 
-local timers = {}
-local mintime
-local function update_timers(delay)
-	mintime = false
-	local sub = 0
-	for index = 1, #timers do
-		index = index - sub
-		local timer = timers[index]
-		timer.time = timer.time - delay
-		if timer.time <= 0 then
-			core.set_last_run_mod(timer.mod_origin)
-			timer.func(unpack(timer.args or {}))
-			table.remove(timers, index)
-			sub = sub + 1
-		elseif mintime then
-			mintime = math.min(mintime, timer.time)
-		else
-			mintime = timer.time
-		end
-	end
-end
+local jobs = {}
+local time = 0.0
+local last = 0.0
 
-local timers_to_add
-local function add_timers()
-	for _, timer in ipairs(timers_to_add) do
-		table.insert(timers, timer)
+core.register_globalstep(function(dtime)
+	local new = core.get_us_time() / 1000000
+	if new > last then
+		time = time + (new - last)
+	else
+		-- rollover, may lose a little bit of time here but
+		-- only 1 tick max, potentially running timers slightly
+		-- too early
+		time = time + new
 	end
-	timers_to_add = false
-end
+	last = new
 
-local delay = 0
-core.register_globalstep(function(dtime)
-	if not mintime then
-		-- abort if no timers are running
+	if #jobs < 1 then
 		return
 	end
-	if timers_to_add then
-		add_timers()
-	end
-	delay = delay + dtime
-	if delay < mintime then
-		return
+
+	-- iterate backwards to assure no table corruption
+	for i = #jobs, 1, -1 do
+		if time >= jobs[i].expire then
+			core.set_last_run_mod(jobs[i].mod_origin)
+			jobs[i].func(unpack(jobs[i].arg))
+			table.remove(jobs, i)
```

Comment:
@kilbith If `jobs[i].func` (i.e. the `after` callback) calls `minetest.after` itself, it would leave a `nil` "hole" in the `jobs` array breaking the whole thing.

(`table.remove` could be localized to save a few lookups, but will probably not yield a big enough performance boost anyway).